### PR TITLE
tests: Migrate watch test to common framework

### PR DESCRIPTION
**Pull request by nic-chen**
Part of https://github.com/etcd-io/etcd/issues/13637

### CODE REVIEW

File: tests/common/watch_test.go
Comment (serathius) [inline tests/common/watch_test.go:16]: Why use different list of cluster setups than default?

File: tests/common/watch_test.go
Comment (serathius) [inline tests/common/watch_test.go:84]:
```suggestion
					wCtx, wCancel := context.WithCancel(ctx)
```

File: tests/common/watch_test.go
Comment (serathius) [inline tests/common/watch_test.go:91]: Does this need to be in separate goroutine? With open watch you should be able to do puts synchronously.

File: tests/common/watch_test.go
Comment (serathius) [inline tests/common/watch_test.go:101]: Is this sleep needed?

Comment (serathius) [reply tests/common/watch_test.go:101]: Please update comment to explain why.

File: tests/common/watch_test.go
Comment (serathius) [inline tests/common/watch_test.go:110]: Move this whole loop that reads KV to separate function.

File: tests/framework/e2e/etcdctl.go
Comment (serathius) [inline tests/framework/e2e/etcdctl.go:602]: This looks suspicious. Why this is a problem for e2e tests and not integration tests? That would mean that etcd client library handles it but etcdctl not?

**Comment by serathius** [comment]
Left some small comments, but overall looks great! Good job.

File: tests/common/watch_test.go
Comment (nic-chen) [reply tests/common/watch_test.go]: sure, updated.

File: tests/common/watch_test.go
Comment (nic-chen) [reply tests/common/watch_test.go]: fixed.

File: tests/common/watch_test.go
Comment (nic-chen) [reply tests/common/watch_test.go]: removed the goroutine.

File: tests/common/watch_test.go
Comment (nic-chen) [reply tests/common/watch_test.go]: removed.

File: tests/common/watch_test.go
Comment (nic-chen) [reply tests/common/watch_test.go]: done.

File: tests/framework/e2e/etcdctl.go
Comment (nic-chen) [reply tests/framework/e2e/etcdctl.go]: updated to comments. please take a look.

File: tests/framework/testutils/helpters.go
Comment (serathius) [inline tests/framework/testutils/helpters.go:57]:
```suggestion
					return kvs, nil
```

File: tests/framework/testutils/helpters.go
Comment (serathius) [inline tests/framework/testutils/helpters.go:50]: No need for this label if we use return instead.

File: tests/framework/testutils/helpters.go
Comment (nic-chen) [reply tests/framework/testutils/helpters.go]: updated.

File: tests/framework/testutils/helpters.go
Comment (nic-chen) [reply tests/framework/testutils/helpters.go]: removed.

File: tests/framework/e2e/etcdctl.go
Comment (nic-chen) [reply tests/framework/e2e/etcdctl.go]: Updated. Thanks for the suggestion. I thought about it but didn't make it happen.

Comment (serathius) [reply tests/framework/e2e/etcdctl.go]: Looks great!

**Review by serathius — APPROVED** [review]

File: tests/framework/e2e/etcdctl.go
Comment (ahrtr) [inline tests/framework/e2e/etcdctl.go:584]: Have you considered to wait for the goroutine to exit when the test case finish?

Comment (nic-chen) [reply tests/framework/e2e/etcdctl.go:584]: hi, @ahrtr 
 the goroutine will exit by context cancel when the test finishes, see: https://github.com/etcd-io/etcd/pull/14345/files#diff-70546f01c62cf8d453ef453d51b64bbdc664e39613453b572120148a77600a08R76

Comment (ahrtr) [reply tests/framework/e2e/etcdctl.go:584]: It's possible (although with low possibility) that the test case finish before the goroutine exit, eventually the test might report goroutine leak errors. Our pipeline fails due to this kind of error from time to time. So for safety, we need to make sure the goroutine exits before the the test case returns/exits.

Comment (serathius) [reply tests/framework/e2e/etcdctl.go:584]: I don't think this is a problem. We should be good as long as we propagate signal to goroutine that causes it to exit. It's up to leak detection to wait for this signal to propagate. If you read the etcd test goroutine leak implementation, it gives test up to 5 seconds to cleanup it's goroutines and explicitly gives them time to execute by calling `runtime.Gosched()`.

Comment (nic-chen) [reply tests/framework/e2e/etcdctl.go:584]: I think we could merge it first. If there is a real problem, I will fix it. What do you think? @serathius @ahrtr

Comment (serathius) [reply tests/framework/e2e/etcdctl.go:584]: I think @ahrtr is worried that the problem will not just imminently show up after merging. Leak goroutine issues like this one are very subtle (for example I cannot reproduce this issue locally at all), so I understand his position.

I still think my argument holds and this should not be a problem, but I would prefer to wait for confirmation from @ahrtr

Comment (ahrtr) [reply tests/framework/e2e/etcdctl.go:584]: > If you read the etcd test goroutine leak implementation, it gives test up to 5 seconds to cleanup its goroutines and uses `runtime.Gosched()` to give goroutine time to execute and exit.

The goroutine leak checker indeed waits up to [1 second](https://github.com/etcd-io/etcd/blob/e363480089783aff0d1c26586f703d31223cef5d/client/pkg/testutil/leak.go#L117) in the `afterTest` check. 

This is just a mitigation solution instead of best practice to me. We should always make sure gracefully shutdown of all threads/goroutines in almost all cases, including production code and test code, unless the cases that we intentionally create.

Previously we also saw a situation that a test case (in test code) or the main goroutine (in production code) already finishes/exits, but a goroutine it creates still prints log, and it isn't allowed by golang. Of course, I do not see such issue in this PR.

I just searched the repo, there are lots of similar cases which create goroutines but do not gracefully shutdown them. It's OK to merge this PR for now. But I'd like to revisit all such cases afterwards in separate PR(s).  WDYT? @serathius @spzala

Comment (serathius) [reply tests/framework/e2e/etcdctl.go:584]: I agree the is a problem worth looking into.

**Comment by serathius** [comment]
Merging as we agreed to follow up with separate PRs.

**Pull request merged**
